File: hackerrank/problem-solving/heap/jesse-and-cookies.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(str => str.trim());

    main();
});

function readLine() {
    return inputString[currentLine++];
}

//<myCode>------------------------------------------------

const Heap = function(compare){
    this._heap = [];
    this._compare = compare;
}

Heap.prototype.swapValues = function(i1,i2){
    const temp = this._heap[i1];
    this._heap[i1] = this._heap[i2];
    this._heap[i2] = temp;
}

Heap.prototype.printHeapArray = function(){
    console.log(this._heap);
}

Heap.prototype.getParentIndex = function(index){
    return parseInt((index-1)/2);
}

Heap.prototype.getRightChildIndex = function(index){
    const rightChildIndex = parseInt((index*2)+2);
    return rightChildIndex<this._heap.length ? rightChildIndex : null;
}

Heap.prototype.getLeftChildIndex = function(index){
    const leftChildIndex = parseInt((index*2)+1);
    return leftChildIndex<this._heap.length ? leftChildIndex : null;
}


Heap.prototype.heapifyUp = function(){
    let currentIndex = this._heap.length - 1;
    while(currentIndex>0){
        const parentIndex = this.getParentIndex(currentIndex);
        if(this._compare(this._heap[currentIndex],this._heap[parentIndex])>0){
            this.swapValues(parentIndex,currentIndex);
            currentIndex = parentIndex;
        } else {
            break;
        }
    }
}

Heap.prototype.heapifyDown = function(){
    let currentIndex = 0;
    while(currentIndex<this._heap.length){
        const leftChildIndex = this.getLeftChildIndex(currentIndex);
        if(leftChildIndex){
            const rightChildIndex = this.getRightChildIndex(currentIndex);
            let candidateIndex = rightChildIndex && this._compare(this._heap[rightChildIndex],this._heap[leftChildIndex])>0 ? rightChildIndex : leftChildIndex;
            if(this._compare(this._heap[candidateIndex],this._heap[currentIndex])>0){
                this.swapValues(candidateIndex,currentIndex);
                currentIndex = candidateIndex;
            } else {
                break;
            }
        } else {
            break;
        }
    }
}

Heap.prototype.getSize = function(){
    return this._heap.length;
}

Heap.prototype.peek = function(){
    return this._heap[0];
}

Heap.prototype.add = function(value){
    this._heap.push(value);
    this.heapifyUp();
}

Heap.prototype.extract = function(){
    const extractedValue = this._heap[0] || null;
    const lastValue = this._heap.pop();
    if(this._heap.length!=0){
        this._heap[0] = lastValue;
        this.heapifyDown();
    }
    return extractedValue;
}
const getNewCookieSweetness = (s1,s2) => parseInt(Math.min(s1,s2) + (2*Math.max(s1,s2)));

/*
 * Complete the cookies function below.
 */
function cookies(k, A) {
    /*
     * Write your code here.
     */
    let stepCount = 0;
    const cookieHeap = new Heap((i,j) => (j-i));

    for(let i=0;i<A.length;++i){
        cookieHeap.add(A[i]);
    }

    while(cookieHeap.getSize()>=2 && cookieHeap.peek()<k){
        ++stepCount;
        let a = cookieHeap.extract();
        let b = cookieHeap.extract();
        cookieHeap.add(getNewCookieSweetness(a,b));
    }
    console.log(`${stepCount} ${cookieHeap.peek()}`);
    return cookieHeap.peek()>=k ? stepCount : -1;
}

//</myCode>------------------------------------------


function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const nk = readLine().split(' ');

    const n = parseInt(nk[0], 10);

    const k = parseInt(nk[1], 10);

    const A = readLine().split(' ').map(ATemp => parseInt(ATemp, 10));

    let result = cookies(k, A);

    ws.write(result + "\n");

    ws.end();
}

File: hackerrank/problem-solving/heap/jesse-and-cookies.java
import java.io.*;
import java.math.*;
import java.text.*;
import java.util.*;
import java.util.regex.*;

//<myCode>---------------
 class MinHeapIntegerComparator implements Comparator<Integer> {
    public int compare(Integer i1, Integer i2){
        return i1.intValue()-i2.intValue();
    }
}
//</myCode>-----------------
public class Solution {

    /*
     * Complete the cookies function below.
     */


    //<myCode>---------------------------

    private static Integer getCombinedCookie(Integer c1, Integer c2){
        return c1+(2*c2);
    }


    static int cookies(int k, int[] A) {
        /*
         * Write your code here.
         */

         PriorityQueue<Integer> cookieHeap = new PriorityQueue<Integer>(new MinHeapIntegerComparator());

         for(int cookieVal : A){
             cookieHeap.add(cookieVal);
         }
        int counter = 0;
         while(cookieHeap.size()>=2 && cookieHeap.peek()<k){
             Integer c1 = cookieHeap.poll();
             Integer c2 = cookieHeap.poll();
            cookieHeap.add(getCombinedCookie(c1,c2));
            ++counter;
         }

        return cookieHeap.peek()<k ? -1 : counter;

    }

    //</myCode>-------------------------

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nk = scanner.nextLine().split(" ");

        int n = Integer.parseInt(nk[0].trim());

        int k = Integer.parseInt(nk[1].trim());

        int[] A = new int[n];

        String[] AItems = scanner.nextLine().split(" ");

        for (int AItr = 0; AItr < n; AItr++) {
            int AItem = Integer.parseInt(AItems[AItr].trim());
            A[AItr] = AItem;
        }

        int result = cookies(k, A);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();
    }
}
